### DISCUSSION TITLE
iPXE loop problem

### DISCUSSION THREAD

**Discussion by mpwyn** [ltsp/ltsp#639] [General]
When we used the version of undionly.kpxe installed by ltsp we got a loop that ended with the error message shown below. We tried downloading the version from boot.ipxe.org as suggested in https://github.com/ltsp/ltsp/discussions/276#discussioncomment-102218, but that didn't work.

We were able to get the system to boot by using an undionly.kpxe borrowed from the LTSP5 system we're trying to replace. To do that, though, we had to make alterations in the LTSP21 setup to get it to work because our old version of undionly.kpxe retrieves the files using http instead of tftp. 

We'd rather not go that way because we'd like to set up our new LTSP21 to match the standard setup as much as possible. Any suggestions? 

![7a998a0c-887d-49ba-8a08-b74bdb53817d](https://user-images.githubusercontent.com/13062574/149508490-884f61d8-06e4-45c3-b321-3f6f5ec34c98.jpg)

**Comment by alkisg** [comment]
You need to describe your setup in more detail.

Are you using a different or external DHCP server, instead of dnsmasq that is automatically configured by LTSP?
If so, which software is it running, and what is its exact configuration, that's causing the loop?
Why do you want http? Where is your http server, is it the same as the LTSP server?
Etc etc, we can't provide instructions without knowing the setup...

**Reply by mpwyn** [reply]
We're using Infoblox. The ltsp server is set to be the next server. The boot file is defined as "undionly.kpxe".

Hm, I wonder if maybe all we need to do is specify a path for undionly.kpxe? I'll try that out.

Please feel free to make any other suggestions, though, if you have any.

The only reason I mentioned the http server is that that was how we have it set up in our LTSP5 installation and we were able to successfully use undionly.kpxe from that installation. We're not necessarily planning on doing it that way with LTSP21 so it's not really important for the current question. 

It was the http part of the equation that caused problems today because we had used a symlink from the Apache area to the LTSP21 area, and it turned out that symlinks don't work. I don't remember off hand all the reasons why I chose to use http so I'd have to look it up in my notes to see if I can figure it out.

**Comment by mpwyn** [comment]
No, it doesn't look like the path is the problem. The tftpd-hpa daemon is running like this:

`/usr/sbin/in.tftpd --listen --user tftp --address 0.0.0.0:69 --secure /srv/tftp/ltsp`

And the undionly is located there:

/srv/tftp/ltsp# ls -la
drwxr-xr-x 5 root root      4096 Jan 14 14:10 .
drwxr-xr-x 3 root nogroup   4096 Jan 31  2019 ..
...
-rw-r--r-- 1 root root     67021 Jan 14 12:20 undionly.kpxe
...

**Comment by alkisg** [comment]
The easiest way would be to disable all the BOOTP/PXE options in your infoblox, and run `ltsp dnsmasq` on the LTSP server, and allow it to be the proxyDHCP server and TFTP server.

You don't need tftpd-hpa at all; `ltsp dnsmasq` automatically configures dnsmasq to be the TFTP server as well.

**Reply by mpwyn** [reply]
Yeah, I can see that that would be the easiest way. My first reaction is to mention that we're working in a university environment and the university computer center doesn't allow individual groups to run their own DHCP server. Then I saw that you wrote "proxyDHCP" so I looked that up and it seems to be different from a normal DHCP server.

I'm not familiar with proxyDHCP. One problem I can imagine we might have is that our LTSP clients are in different subnets so I'm thinking the LTSP/dnsmasq server won't see the requests from clients outside of its own subnet. Is that the case? Is there a solution for that?

**Comment by alkisg** [comment]
The upstream documentation for breaking the iPXE loop is there: https://ipxe.org/howto/chainloading#breaking_the_infinite_loop
The LTSP configuration for dnsmasq is there: https://github.com/ltsp/ltsp/blob/main/ltsp/server/dnsmasq/ltsp-dnsmasq.conf
A similar configuration for isc-dhcp is there: https://ltsp.org/advanced/isc-dhcp-server/

All these have "if / then" commands that make sure that BIOS clients get different files from UEFI clients, and that ltsp.ipxe is sent after iPXE is loaded, and that Raspberry Pi clients get their own configuration etc. It's not a plain "boot filename" anymore.

You need to read these and then do the same thing in infoblox, which I've never used so I can't help with the details.

Regarding proxyDHCP in different subnets, you'd need a proxyDHCP there too, or at least a relay, which is harder to set up.

Finally, you can compile your own snponly.efi and undionly.kpxe that break the loop on their own because they hardcode the boot filename to ltsp/ltsp.ipxe. You can ask for help on the #ipxe Matrix/IRC channels if you choose to do that.

**Reply by mpwyn** [reply]
Thanks for the pointers, Alkis! We'll look at them try to figure out the best way to do things.
